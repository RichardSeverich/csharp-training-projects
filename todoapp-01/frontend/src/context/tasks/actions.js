import { ActionTypes } from '../../enum';
import { editStatus } from 'helpers/actions';
import noop from 'helpers/noop';
import { deleteData, updateStatus, updateData, postDataWithToken } from 'api/api';
import filter from 'helpers/filter';
import processTask from 'helpers/process';
import { getToken } from 'helpers/helper';

/**
 * @param {Function} dispatch function.
 * @param {object} payload is a task.
 */
function OnTaskAdd(dispatch, payload) {
	const newTask = processTask(payload);
	postDataWithToken('tasks/new-task', newTask, getToken()).catch((error) => console.error(error));
}

/**
 * @param {Function} dispatch function.
 * @param {object[]} payload list objects.
 */
function OnTaskChange(dispatch, payload) {
	let [id, task] = payload;
	const newTask = processTask(task);
	updateData(`tasks/${id}`, newTask, getToken()).catch((error) => console.error(error));
}

/**
 * @param {Function} dispatch function.
 * @param {string} payload task uuid.
 */
function OnTaskDelete(dispatch, payload) {
	deleteData(`tasks/${payload}`, getToken()).catch((error) => console.error(error));
}

/**
 * @param {Function} dispatch function.
 * @param {object} payload new content.
 */
function OnTaskStatusChange(dispatch, payload) {
	let [id, body] = editStatus(payload);
	updateStatus(`tasks/${id}`, body, getToken()).catch((error) => console.error(error));
}

/**
 * @param {Function} dispatch function.
 * @param {object} payload load parameters.
 */
function OnLoad(dispatch, payload) {
	dispatch({ type: ActionTypes.TaskLoadSuccess, payload: true });
	setTimeout(
		() =>
			postDataWithToken('tasks', payload, getToken())
				.then((response) => response.json())
				.then((response) => {
					const { currentPage } = response;
					let tasks = response.data ? response.data : [];
					let tasksByStatus = filter('status', tasks);
					const currentPageS = currentPage <= response.totalPages ? currentPage : 1;
					dispatch({ type: ActionTypes.TaskLoadSuccess, payload: false });
					dispatch({
						type: ActionTypes.TaskLoad,
						payload: {
							...response,
							currentPage: currentPageS,
							data: { all: tasks, byStatus: tasksByStatus },
						},
					});
				})
				.catch((error) => {
					OnTaskLoading(dispatch, false);
					console.error(error);
				}),
		1000
	);
}

/**
 * @param {Function} dispatch function.
 * @param {boolean} payload is a state.
 */
function OnTaskLoading(dispatch, payload) {
	dispatch({ type: ActionTypes.TaskLoadSuccess, payload });
}

/**
 * @param {Function} dispatch function.
 * @param {object} payload new page content.
 */
function OnPageChange(dispatch, payload) {
	dispatch({ type: ActionTypes.PageChanged, payload });
}

/**
 * @param {Function} dispatch function.
 * @param {object} payload new search content.
 */
function OnSearchChange(dispatch, payload) {
	dispatch({ type: ActionTypes.SearchChanged, payload });
}

/**
 * @param {Function} dispatch function.
 * @returns {object} callback actions.
 */
export default function ActionFactory(dispatch = noop) {
	return {
		OnTaskAdd: (payload) => OnTaskAdd(dispatch, payload),
		OnTaskChange: (payload) => OnTaskChange(dispatch, payload),
		OnTaskDelete: (payload) => OnTaskDelete(dispatch, payload),
		OnTaskStatusChange: (payload) => OnTaskStatusChange(dispatch, payload),
		OnTaskLoading: (payload) => OnTaskLoading(dispatch, payload),
		OnLoad: (payload) => OnLoad(dispatch, payload),
		OnPageChange: (payload) => OnPageChange(dispatch, payload),
		OnSearchChange: (payload) => OnSearchChange(dispatch, payload),
	};
}
