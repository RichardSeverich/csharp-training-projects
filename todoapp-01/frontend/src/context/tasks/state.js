import { ActionTypes } from '../../enum';

/**
 * @param {object} props props received.
 * @returns {object} initial state.
 */
export function State(props) {
	return {
		...props,
		search: {
			description: '',
			priority: '',
			project: '',
			tags: [],
			entry: 'any',
			status: ['N!Deleted', ''],
		},
		currentPage: 1,
		pageSize: 5,
		totalPages: 1,
		data: { all: [], byStatus: [] },
		loading: false,
	};
}

/**
 * @param {object} state current state.
 * @param {object} action current action.
 * @returns {object} new state.
 */
export function Reducer(state, action) {
	const { type, payload } = action;

	switch (type) {
		case ActionTypes.TaskAdd:
			return { ...state, data: { ...payload } };
		case ActionTypes.TaskChange:
			return { ...state, data: { ...payload } };
		case ActionTypes.TaskDelete:
			return { ...state, data: { ...payload } };
		case ActionTypes.TaskLoad:
			return { ...state, ...payload, loading: false };
		case ActionTypes.TaskStatusChange:
			return { ...state, data: { ...payload } };
		case ActionTypes.LoadingChange:
			return { ...state, loading: payload };
		case ActionTypes.TaskLoadSuccess:
			return { ...state, loading: payload };
		case ActionTypes.PageChanged:
			return { ...state, ...payload };
		case ActionTypes.SearchChanged:
			return { ...state, search: { ...payload } };
		default:
			return state;
	}
}
