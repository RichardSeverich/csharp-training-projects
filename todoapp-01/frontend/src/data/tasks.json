{
    "all": [
        {"id": "8123", "description":"First task", "priority": "H", "due": "12/01/2021", "entry": "11/01/2021", "status": "Pending", "tags":[]},
        {"id": "2141", "description":"Second task", "priority": "M", "due": "12/01/2021", "entry": "12/01/2021", "status": "Pending", "tags":[]}, 
        {"id": "5612", "description":"Third task", "priority": "L", "due": "12/01/2021", "entry": "13/01/2021", "status": "Pending", "tags":[]},
        {"id": "1254", "description":"First task", "priority": "H", "due": "12/01/2021", "entry": "14/01/2021", "status": "In progress", "tags":[]},
        {"id": "5678", "description":"Second task", "priority": "M", "due": "12/01/2021", "entry": "15/01/2021", "status": "In progress", "tags":[]},
        {"id": "9012", "description":"Third task", "priority": "L", "due": "12/01/2021", "entry": "16/01/2021", "status": "In progress", "tags":[]},
        {"id": "1234", "description":"First task", "priority": "H", "due": "12/01/2021", "entry": "17/01/2021", "status": "In progress", "tags":[]},
        {"id":"5108", "description":"Second task", "priority": "M", "due": "12/01/2021", "entry": "18/01/2021", "status": "Completed", "tags":[]},
        {"id":"5163", "description":"Third task", "priority": "L", "due": "12/01/2021", "entry": "19/01/2021", "status": "Completed", "tags":[]}
    ],
    "byStatus": [
        {
            "status": "Pending",
            "tasks": [
                {"id": "8123", "description":"First task", "priority": "H", "due": "12/01/2021", "entry": "11/01/2021", "status": "Pending", "tags":[]},
                {"id": "2141", "description":"Second task", "priority": "M", "due": "12/01/2021", "entry": "12/01/2021", "status": "Pending", "tags":[]}, 
                {"id": "5612", "description":"Third task", "priority": "L", "due": "12/01/2021", "entry": "13/01/2021", "status": "Pending", "tags":[]}    
            ]
        },
        {
            "status": "In progress",
            "tasks": [
                {"id": "1254", "description":"First task", "priority": "H", "due": "12/01/2021", "entry": "14/01/2021", "status": "In progress", "tags":[]},
                {"id": "5678", "description":"Second task", "priority": "M", "due": "12/01/2021", "entry": "15/01/2021", "status": "In progress", "tags":[]},
                {"id": "9012", "description":"Third task", "priority": "L", "due": "12/01/2021", "entry": "16/01/2021", "status": "In progress", "tags":[]},
                {"id": "1234", "description":"First task", "priority": "H", "due": "12/01/2021", "entry": "17/01/2021", "status": "In progress", "tags":[]}
            ]
        },
        {
            "status": "Completed",
            "tasks": [
                {"id":"5108", "description":"Second task", "priority": "M", "due": "12/01/2021", "entry": "18/01/2021", "status": "Completed", "tags":[]},
                {"id":"5163", "description":"Third task", "priority": "L", "due": "12/01/2021", "entry": "19/01/2021", "status": "Completed", "tags":[]}
            ]
        }
    ]
}