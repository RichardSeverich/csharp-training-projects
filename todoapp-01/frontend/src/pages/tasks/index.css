.container {
	width: 90%;
	max-width: 1350px;
	margin: 0 auto;
}

.modal .header {
	background-color: #0f091a !important;
	color: #f0ffff !important;
}
