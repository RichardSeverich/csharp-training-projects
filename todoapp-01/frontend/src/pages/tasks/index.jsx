import React, { useState, useEffect } from 'react';
import { Form, Button, Icon } from 'semantic-ui-react';
import { Link } from 'react-router-dom';
import './index.css';
import TasksView from './views/task';
import Header from 'components/header';
import MenuPages from 'components/menu';
import { useDataProvider } from 'context/tasks/tasks';
import { useAppProvider } from 'context/application/app';
import Pagination from './components/pagination';
import TaskFilter from './components/task_filter';
import { processSearch } from 'helpers/process';

const options = [
	{ key: 'a', text: 'Show all', value: 'all' },
	{ key: 's', text: 'Show by Status', value: 'status' },
];

/**
 * @returns {object} Returns page of tasks.
 */
function Tasks(props) {
	const [, actions] = useDataProvider();
	const [stateApp] = useAppProvider();
	const [state] = useDataProvider();
	const [show, setShow] = useState('all');

	useEffect(() => {
		const searchValue = processSearch(state.search);
		const payload = {
			pageNumber: state.currentPage,
			pageSize: state.pageSize,
			...searchValue, 
		};
		actions.OnLoad(payload);
	}, [stateApp, state.pageSize, state.currentPage, state.search])

	const handleChange = (e, { value }) => {
		setShow(value);
	};

	return (
		<div className="main">
			<Header />
			<MenuPages />
			<TaskFilter search={state.search} />
			<Button icon to={'/tasks/add-task'} as={Link} labelPosition="left" color="grey">
				<Icon name="plus" />
				Add Task
			</Button>
			<Form>
				<Form.Select
					options={options}
					placeholder="Show by..."
					name="show"
					value={show}
					onChange={handleChange}
				/>
			</Form>
			<TasksView show={show} />
			<Pagination currentPage={state.currentPage} pageSize={state.pageSize} totalPages={state.totalPages} />
		</div>
	);
}

export default Tasks;
