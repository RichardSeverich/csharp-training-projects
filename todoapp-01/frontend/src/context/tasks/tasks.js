import React, { createContext, useContext, useMemo, useReducer, useEffect } from 'react';
import { State, Reducer } from './state';
import ActionFactory from './actions';

export const TaskContext = createContext();
export const TaskDispatch = createContext();

/**
 * @param {object} props props received.
 * @returns {React.Component} something.
 */
export function TaskProvider(props) {
	const { children } = props;
	const [value, dispatch] = useReducer(Reducer, props, State);
	const actions = useMemo(() => ActionFactory(dispatch), [dispatch]);

	useEffect(() => {
		/*const payload = {
			pageNumber: value.currentPage,
			pageSize: value.pageSize,
			...value.search, 
		};
		actions.OnLoad(payload);*/
	}, [actions]);

	return (
		<TaskContext.Provider value={value}>
			<TaskDispatch.Provider value={actions}>{children}</TaskDispatch.Provider>
		</TaskContext.Provider>
	);
}

/**
 * @returns {object[]} state and actions.
 */
export function useDataProvider() {
	const state = useContext(TaskContext);
	const actions = useContext(TaskDispatch);

	return [state, actions];
}
