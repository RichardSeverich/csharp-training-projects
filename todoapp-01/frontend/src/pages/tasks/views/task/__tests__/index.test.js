import React from 'react';
import TaskView from '../.';
import { render } from '@testing-library/react';
import { TaskProvider } from 'context/tasks/tasks';

describe('pages/tasks/views/task', () => {
	it('should be instance of', () => {
		const { container } = render(
			<TaskProvider>
				<TaskView />
			</TaskProvider>
		);
		expect(container.firstChild).toBeInstanceOf(HTMLDivElement);
	});

	it('should render correctly', () => {
		const { container } = render(
			<TaskProvider>
				<TaskView />
			</TaskProvider>
		);
		expect(container.firstChild).toMatchSnapshot();
	});

	it('children should have instance of', () => {
		const { container } = render(
			<TaskProvider>
				<TaskView />
			</TaskProvider>
		);
		expect(container.firstChild.firstChild).toBeInstanceOf(HTMLDivElement);
	});
});
