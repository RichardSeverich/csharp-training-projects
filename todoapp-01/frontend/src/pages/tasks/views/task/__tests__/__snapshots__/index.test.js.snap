// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/tasks/views/task should render correctly 1`] = `
<div
  class="ui segment"
>
  <div
    class="accordion ui styled"
  >
    <h3>
      There is nothing.
    </h3>
  </div>
</div>
`;
